xdsl/ir.py

```diff
@@ -972,25 +1092,95 @@ def _attach_op(self, operation: Operation) -> None:
     @property
     def is_empty(self) -> bool:
         """Returns `True` if there are no operations in this block."""
-        return not len(self.ops)
+        return self._first_op is None
 
     @property
     def first_op(self) -> Operation | None:
         """The first operation in this block."""
-        return self.ops[0] if len(self.ops) else None
+        return self._first_op
 
     @property
     def last_op(self) -> Operation | None:
         """The last operation in this block."""
-        return self.ops[-1] if len(self.ops) else None
+        return self._last_op
+
+    def _op_at_index(self,
+                     index: int,
+                     message_lambda: Callable[[], str] | None = None
+                     ) -> Operation | None:
+        """
+        Returns the operation at a given index, supporting positive and negative indices.
+        If abs(index) == num_ops, return None.
+        if abs(index) > num_ops, raise ValueError, optionally with given message.
+        """
+        if index < 0:
+            # Enumerate backwards
+            op = self.last_op
+            while index < 0:
+                if op is None:
+                    break
+                op = op.prev_op
+                index += 1
+            else:
+                return op
+        else:
+            # Enumerate forwards
+            op = self.first_op
+            while index:
+                if op is None:
+                    break
+                op = op.next_op
+                index -= 1
+            else:
+                return op
+        if message_lambda is None:
+            message = f"Invalid operation index {index} for block with {len(self.ops)} ops"
+        else:
+            message = message_lambda()
+        raise ValueError(message)
+
+    def op_at_index(self, index: int) -> Operation:
```

Comment:
I'd love to delete this soon, but it's still useful for all the methods that use op index

Reply:
Can we just for the meantime use something like `list(self.ops)[index]`?
Since we will remove it, let's have a simple implementation rather than a complex optimized one?

Reply:
IDK, I feel like it's a motivator to get rid of all calls to this function for the time being